### DISCUSSION TITLE
How to compare a run with a high CPU usage versus a run with low CPU usage

### DISCUSSION THREAD

**Discussion by sregger** [async-profiler/async-profiler#534] [Q&A] [answered]
I only became aware of the async-profiler a day ago. I'm a novice with this tool and inexperienced with profiling in general. This question may be basic and/or incorrect.

I'm trying to diagnose a service which I've noticed, during load testing, generates an inconsistent CPU pattern. Namely in some runs it maxes out the CPU while in others the CPU is fine. This occurs with the same traffic volume and traffic consistency.

I've been unable to determine what causes it. Using the async-profiler I captured a flame graph for both scenarios. However they look the same. I can't use them to check which areas are doing more work. Likely because the work balance is the same, though the overall CPU usage is different.

My question is if there's a trick I'm missing? Is it possible to visualise absolute CPU usage in order to determine the difference between runs? i.e. The areas that are busier. Or are flame graphs used only for relative results?

**Comment by apangin** [comment, marked as answer]
You're right - two flame graphs will look similar even if they have different number of samples, but the proportion of methods remains the same. However, the popup hint on each frame will show the actual number of samples, which will differ in this case:

![num-samples](https://user-images.githubusercontent.com/1749416/150881430-f2211d75-c324-4c02-90c1-fb47e2bcfd4d.png)

You can visualize CPU load, if you record the profile in .jfr format. This is how it looks in JMC, for example:

![cpu-graph](https://user-images.githubusercontent.com/1749416/150881746-4831cc88-39ea-41db-bc13-b503e3295af1.png)

Note that you can always produce a flame graph from .jfr using `jfr2flame` converter bundled with async-profiler.

Alternatively, you can compare two flame graphs using IntelliJ IDEA ["Compare With Baseline"](https://blog.jetbrains.com/idea/2021/10/5-best-features-for-profiling-java-code/#jfr_snapshots_comparison) feature.

![flame-diff](https://user-images.githubusercontent.com/1749416/150882549-e5f5799d-65a6-4b2b-b2d0-ea8fe427c3f8.png)

**Reply by sregger** [reply]
> However, the popup hint on each frame will show the actual number of samples, which will differ in this case:

I did not realise that samples could be used to determine overall CPU usage. I thought a sample was captured every 10 ms and the final figure was related to how long the profiler was run? e.g. If the profiler was run for 5 minutes and 10 minutes there would be more samples in the 10 minute run.

> if you record the profile in .jfr format

This is on the top of my agenda. But, in my setup, I'm having trouble capturing `jfr` profiles.

> Compare With Baseline

That's very useful. Hopefully once I solve the `jfr` profile issue I can use it.

**Reply by apangin** [reply]
> I thought a sample was captured every 10 ms

Every 10 ms of **CPU time**. That is, when CPU is idle, no samples are collected. See #466 for details on counting samples.

**Comment by krzysztofslusarski** [comment]
There is also a possibility that your application sleeps/waits for a lock/waits for some io/socket response.

Such cases you won't see at the CPU profiling, you need to use wall mode.

**Reply by sregger** [reply]
That's useful advice as this service does rely heavily on a websocket. I'll look into wall mode.

**Reply by toktarev** [reply]
@krzysztofslusarski thanks for your advice.
I experience that same problem as @sregger.

I can't find the bottleneck using samples - because it happens during long socket's awaitings.
